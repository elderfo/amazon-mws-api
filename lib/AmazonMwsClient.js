const qs = require("querystring");
const crypto = require('crypto');
const xml2js = require('xml2js');
const request = require('request');
const _ = require('underscore');
const tls = require('tls');

/**
 * Constructor for the main MWS client interface used to make api calls and
 * various data structures to encapsulate MWS requests, definitions, etc.
 *
 * @param {String} accessKeyId     Id for your secret Access Key (required)
 * @param {String} secretAccessKey Secret Access Key provided by Amazon (required)
 * @param {String} merchantId      Aka SellerId, provided by Amazon (required)
 * @param {Object} options         Additional configuration options for this instance
 */
function AmazonMwsClient(accessKeyId, secretAccessKey, merchantId, options) {
  options = options || {};

  const createCredentials = tls.createSecureContext || crypto.createCredentials;
  this.host = options.host || 'mws.amazonservices.com';
  this.creds = createCredentials(options.creds || {});
  this.appName = options.appName || 'mws-js';
  this.appVersion = options.appVersion || '0.1.0';
  this.appLanguage = options.appLanguage || 'JavaScript';
  this.accessKeyId = accessKeyId || null;
  this.secretAccessKey = secretAccessKey || null;
  this.merchantId = merchantId || null;
  this.authToken = options.authToken;
  this.explicitArray = options.explicitArray || true;
}

/**
 * The method used to invoke calls against MWS Endpoints. Recommended usage is
 * through the invoke wrapper method when the api call you're invoking has a
 * request defined in one of the submodules. However, you can use call() manually
 * when a lower level of control is necessary (custom or new requests, for example).
 *
 * @param  {Object}   api      Settings object unique to each API submodule
 * @param  {String}   action   Api `Action`, such as GetServiceStatus or GetOrder
 * @param  {Object}   query    Any parameters belonging to the current action
 * @return Promise
 */
AmazonMwsClient.prototype.call = function (api, action, query) {
  if (this.secretAccessKey == null || this.accessKeyId == null || this.merchantId == null) {
    throw "accessKeyId, secretAccessKey, and merchantId must be set";
  }

  let requestOpts = {
    method: 'POST',
    uri: 'https://' + this.host + api.path
  };

  // Check if we're dealing with a file (such as a feed) upload
  if (api.upload) {
    requestOpts.body = query._BODY_;
    query._FORMAT_ = 'application/x-www-form-urlencoded';
    requestOpts.headers = {
      'Content-Type': query._FORMAT_,
      'Content-MD5': crypto.createHash('md5').update(query._BODY_).digest('base64')
    };
    delete query._BODY_;
    delete query._FORMAT_;
  }

  // Add required parameters and sign the query
  query['Action'] = action;
  query['Version'] = api.version;
  query["Timestamp"] = new Date().toISOString();
  query["AWSAccessKeyId"] = this.accessKeyId;

  if (this.authToken) {
    query["MWSAuthToken"] = this.authToken;
  }

  if (api.legacy) {
    query['Merchant'] = this.merchantId;
  } else {
    query['SellerId'] = this.merchantId;
  }

  query = this.sign(api.path, query, this.host);

  if (!api.upload) {
    requestOpts.form = query;
  } else {
    requestOpts.qs = query;
  }

  return new Promise(function (resolve, reject) {
    request(requestOpts, function (err, response, data) {
      if (err) {
        reject(err);
      } else {
        if (data.slice(0, 5) == '<?xml') {
          const parser = new xml2js.Parser({explicitArray: this.explicitArray});
          parser.parseString(data, function (err, result) {
            // Throw an error if there was a problem reported
            if (err != null) {
              reject(new Error(err.Code + ": " + err.Message));
            } else {
              resolve(result);
            }
          });
        } else {
          resolve(data);
        }
      }
    });
  });
};

/**
 * Calculates the HmacSHA256 signature and appends it with additional signature
 * parameters to the provided query object.
 *
 * @param  {String} path  Path of API call (used to build the string to sign)
 * @param  {Object} query Any non-signature parameters that will be sent
 * @return {Object}       Finalized object used to build query string of request
 */
AmazonMwsClient.prototype.sign = function (path, query, host) {
  // Configure the query signature method/version
  query["SignatureMethod"] = "HmacSHA256";
  query["SignatureVersion"] = "2";

  // Copy query keys, sort them, then copy over the values
  const sorted = _.reduce(_.keys(query).sort(), function (m, k) {
    m[k] = query[k];
    return m;
  }, {});

  const stringToSign = ["POST", host, path, qs.stringify(sorted)].join("\n");

  query['Signature'] = crypto.createHmac("sha256", this.secretAccessKey).update(stringToSign, 'utf8').digest("base64");

  return query;
};

/**
 * Suggested method for invoking a pre-defined mws request object.
 *
 * @param  {Object}   request  An instance of AmazonMwsRequest with params, etc.
 * @return Promise
 */
AmazonMwsClient.prototype.invoke = function (request) {
  const _this = this;
  return request.query().then(function (q) {
    return _this.call(request.api, request.action, q);
  });
};

export default AmazonMwsClient;
