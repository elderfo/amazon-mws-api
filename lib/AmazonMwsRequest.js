/**
 * Constructor for general MWS request objects, wrapped by api submodules to keep
 * things DRY, yet familiar despite whichever api is being implemented.
 *
 * @param {Object} options Settings to apply to new request instance.
 */

export default class AmazonMwsRequest {
  constructor({
    path = '/',
    version = '2009-01-01',
    legacy = false,
    upload,
    action = 'GetServiceStatus',
    params = {},
  }) {
    this.api = {
      path,
      version,
      legacy,
      upload,
    };
    this.action = action;
    this.params = { ...params };
    this.paramsMap = {};

    if (Object.keys(this.params).length > 0) {
      for (const paramName in this.params) {
        const realName = this.params[paramName].name;
        if (paramName !== this.params[paramName].name) {
          this.paramsMap[paramName] = realName;
          this.params[realName] = this.params[paramName];
          delete this.params[paramName];
        }
      }
    }
  }

  setMultiple(conf) {
    Object.keys(conf).forEach((key) => {
      this.set(key, conf[key]);
    });

    return this;
  }

  /**
   * Handles the casting, renaming, and setting of individual request params.
   *
   * @param {String} param Key of parameter (not ALWAYS the same as the param name!)
   * @param {Mixed} value Value to assign to parameter
   * @return {Object} Current instance to allow function chaining
   */
  set(param, value) {
    if (param instanceof Object && (value === null || value === undefined)) {
      return this.setMultiple(param);
    } else if (value !== null && value !== undefined) {
      const self = this;

      if (this.paramsMap.hasOwnProperty(param)) {
        param = this.paramsMap[param];
      }

      const p = this.params[param],
        v = p.value = {};
      // Handles the actual setting based on type
      const setListValue = function setListValue(name, val) {
        if (p.type == 'Timestamp') {
          v[name] = val.toISOString();
        } else if (p.type == 'Boolean') {
          v[name] = val ? 'true' : 'false';
        } else {
          v[name] = val;
        }
      };

      // Handles the actual setting based on type
      const setValue = function setValue(name, val) {
        if (p.type == 'Timestamp') {
          self.params[name].value = val.toISOString();
        } else if (p.type == 'Boolean') {
          self.params[name].value = val ? 'true' : 'false';
        } else {
          self.params[name].value = val;
        }
      };

      // Lists need to be sequentially numbered and we take care of that here
      if (p.list) {
        let i = 0;
        if (typeof value === 'string' || typeof value === 'number') {
          setListValue(`${p.name}.1`, value);
        }
        if (typeof value === 'object') {
          if (Array.isArray(value)) {
            for (i = value.length - 1; i >= 0; i--) {
              setListValue(`${p.name}.${i + 1}`, value[i]);
            }
          } else {
            for (const key in value) {
              setListValue(`${p.name}.${++i}`, value[key]);
            }
          }
        }
      } else {
        setValue(p.name, value);
      }
    }
    return this;
  }

  /**
   * Builds a query object and checks for required parameters.
   *
   * @return {Object} KvP's of all provided parameters (used by invoke())
   */
  query() {
    const _this = this;
    return new Promise((resolve, reject) => {
      let workingQuery = {};
      const missing = [];
      for (const param in _this.params) {
        const value = _this.params[param].value;
        const name = _this.params[param].name;
        const complex = _this.params[param].type === 'Complex';
        const list = _this.params[param].list;
        const required = _this.params[param].required;

        if (value !== undefined && value !== null) {
          if (complex) {
            workingQuery = value.appendTo(workingQuery);
          } else if (list) {
            for (const p in value) {
              workingQuery[p] = value[p];
            }
          } else {
            workingQuery[name] = value;
          }
        } else if (_this.params[param].required === true) {
          missing.push(name);
        }
      }
      if (missing.length > 0) reject(new Error(`ERROR: Missing required parameter(s): ${missing.join(',')}!`));
      else resolve(workingQuery);
    });
  }
}
