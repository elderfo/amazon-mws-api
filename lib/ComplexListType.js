/**
 * Complex List helper object. Once initialized, you should set
 * an add(args) method which pushes a new complex object to members.
 *
 * @param {String} name Name of Complex Type (including .member or subtype)
 */
export default class ComplexListType {
  constructor(name) {
    this.pre = name;
    this.members = [];
  }

  /**
   * Appends each member object as a complex list item
   * @param  {Object} query Query object to append to
   * @return {Object}       query
   */
  appendTo(query) {
    const workingQuery = { ...query };

    this.members.forEach((member, idx) => {
      Object.keys(member).forEach((key) => {
        workingQuery[`${this.pre}.${idx + 1}.${key}`] = member[key];
      });
    });

    return workingQuery;
  }
}
