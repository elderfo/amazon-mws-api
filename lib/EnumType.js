
/**
 * Contructor for objects used to represent enumeration states. Useful
 * when you need to make programmatic updates to an enumerated data type or
 * wish to encapsulate enum states in a handy, re-usable variable.
 *
 * @param {Array} choices An array of any possible values (choices)
 */
export default class EnumType {
  constructor(choices) {
    Object.keys(choices).forEach((key) => {
      this[choices[key]] = false;
    });

    this._choices = choices;
  }

  /**
   * Enable one or more choices (accepts a variable number of arguments)
   * @return {Object} Current instance of EnumType for chaining
   */
  enable(...args) {
    Object.keys(args).forEach((key) => {
      this[args[key]] = true;
    });
    return this;
  }

  /**
   * Disable one or more choices (accepts a variable number of arguments)
   * @return {Object} Current instance of EnumType for chaining
   */
  disable(...args) {
    Object.keys(args).forEach((key) => {
      this[args[key]] = false;
    });
    return this;
  }

  /**
   * Toggles one or more choices (accepts a variable number of arguments)
   * @return {Object} Current instance of EnumType for chaining
   */
  toggle(...args) {
    Object.keys(args).forEach((key) => {
      this[args[key]] = !this[args[key]];
    });
    return this;
  }

  /**
   * Return all possible values without regard to current state
   * @return {Array} Choices passed to EnumType constructor
   */
  all() {
    return this._choices;
  }

  /**
   * Return all enabled choices as an array (used to set list params, usually)
   * @return {Array} Choice values for each choice set to true
   */
  values() {
    return Object.keys(this._choices)
      .filter(choice => this[this._choices[choice]] === true)
      .map(choice => this._choices[choice]);
  }
}

