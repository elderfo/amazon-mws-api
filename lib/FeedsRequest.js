/**
 * Feeds API requests and definitions for Amazon's MWS web services.
 * For information on using, please see examples folder.
 *
 * @author Robert Saunders
 */
import AmazonMwsRequest from './AmazonMwsRequest';
import EnumType from './EnumType';

/**
 * Construct a Feeds API request for mws.Client.invoke()
 *
 * @param {String} action Action parameter of request
 * @param {Object} params Schemas for all supported parameters
 * @param {Boolean} isUpload Flag denoting upload
 */
export default class FeedsRequest extends AmazonMwsRequest {
  constructor(action, params, isUpload) {
    const opts = {
      name: 'Feeds',
      group: 'Feeds',
      path: '/',
      upload: isUpload,
      version: '2009-01-01',
      legacy: true,
      action,
      // This next field is the just the SPECIFICATION (schema) of the parameters.
      // There are no VALUES for these parameters yet.
      params,
    };
    super(opts);
  }
}

/**
 * Objects to represent enum collections used by some request(s)
 * @type {Object}
 */
const enums = exports.enums = {

  FeedProcessingStatuses() {
    return new EnumType(['_SUBMITTED_', '_IN_PROGRESS_', '_CANCELLED_', '_DONE_']);
  },

  FeedTypes() {
    return new EnumType([
      '_POST_PRODUCT_DATA_', '_POST_PRODUCT_RELATIONSHIP_DATA_', '_POST_ITEM_DATA_', '_POST_PRODUCT_OVERRIDES_DATA_', '_POST_PRODUCT_IMAGE_DATA_',
      '_POST_PRODUCT_PRICING_DATA_', '_POST_INVENTORY_AVAILABILITY_DATA_', '_POST_ORDER_ACKNOWLEDGEMENT_DATA_', '_POST_ORDER_FULFILLMENT_DATA_',
      '_POST_FULFILLMENT_ORDER_REQUEST_DATA_', '_POST_FULFILLMENT_ORDER_CANCELLATION', '_POST_PAYMENT_ADJUSTMENT_DATA_', '_POST_INVOICE_CONFIRMATION_DATA_',
      '_POST_FLAT_FILE_LISTINGS_DATA_', '_POST_FLAT_FILE_ORDER_ACKNOWLEDGEMENT_DATA_', '_POST_FLAT_FILE_FULFILLMENT_DATA_',
      '_POST_FLAT_FILE_FBA_CREATE_INBOUND_SHIPMENT_', '_POST_FLAT_FILE_FBA_UPDATE_INBOUND_SHIPMENT_', '_POST_FLAT_FILE_PAYMENT_ADJUSTMENT_DATA_',
      '_POST_FLAT_FILE_INVOICE_CONFIRMATION_DATA_', '_POST_FLAT_FILE_INVLOADER_DATA_', '_POST_FLAT_FILE_CONVERGENCE_LISTINGS_DATA_',
      '_POST_FLAT_FILE_BOOKLOADER_DATA_', '_POST_FLAT_FILE_LISTINGS_DATA_', '_POST_FLAT_FILE_PRICEANDQUANTITYONLY', '_POST_UIEE_BOOKLOADER_DATA_',
    ]);
  },

};

/**
 * A collection of currently supported request constructors. Once created and
 * configured, the returned requests can be passed to an mws client `invoke` call
 * @type {Object}
 */
const calls = exports.requests = {

  CancelFeedSubmissions() {
    return new FeedsRequest('CancelFeedSubmissions', {
      FeedSubmissionIds: { name: 'FeedSubmissionIdList.Id', list: true, required: false },
      FeedTypes: { name: 'FeedTypeList.Type', list: true },
      SubmittdFrom: { name: 'SubmittedFromDate', type: 'Timestamp' },
      SubmittedTo: { name: 'SubmittedToDate', type: 'Timestamp' },
    });
  },

  GetFeedSubmissionList() {
    return new FeedsRequest('GetFeedSubmissionList', {
      FeedSubmissionIds: { name: 'FeedSubmissionIdList.Id', list: true, required: false },
      MaxCount: { name: 'MaxCount' },
      FeedTypes: { name: 'FeedTypeList.Type', list: true },
      FeedProcessingStatuses: { name: 'FeedProcessingStatusList.Status', list: true, type: 'bde.FeedProcessingStatuses' },
      SubmittedFrom: { name: 'SubmittedFromDate', type: 'Timestamp' },
      SubmittedTo: { name: 'SubmittedToDate', type: 'Timestamp' },
    });
  },

  GetFeedSubmissionListByNextToken() {
    return new FeedsRequest('GetFeedSubmissionListByNextToken', {
      NextToken: { name: 'NextToken', required: true },
    });
  },

  GetFeedSubmissionCount() {
    return new FeedsRequest('GetFeedSubmissionCount', {
      FeedTypes: { name: 'FeedTypeList.Type', list: true },
      FeedProcessingStatuses: { name: 'FeedProcessingStatusList.Status', list: true, type: 'bde.FeedProcessingStatuses' },
      SubmittedFrom: { name: 'SubmittedFromDate', type: 'Timestamp' },
      SubmittedTo: { name: 'SubmittedToDate', type: 'Timestamp' },
    });
  },

  GetFeedSubmissionResult() {
    return new FeedsRequest('GetFeedSubmissionResult', {
      FeedSubmissionId: { name: 'FeedSubmissionId', required: true },
    });
  },

  SubmitFeed() {
    return new FeedsRequest('SubmitFeed',
      // schema:
      {
        FeedContents: { name: '_BODY_', required: true },
        FeedType: { name: 'FeedType', required: true },
        MarketplaceIds: { name: 'MarketplaceIdList.Id', list: true, required: false },
        PurgeAndReplace: { name: 'PurgeAndReplace', required: false, type: 'Boolean' },
      }, true);
  },

};
