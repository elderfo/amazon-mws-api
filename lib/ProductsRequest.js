/**
 * Products API requests and definitions for Amazon's MWS web services.
 * For information on using, please see examples folder.
 *
 * @author Robert Saunders
 */
import AmazonMwsRequest from './AmazonMwsRequest';
import EnumType from './EnumType';

/**
 * Construct a Products API request for using with mws.Client.invoke()
 *
 * @param {String} action Action parameter of request
 * @param {Object} params Schemas for all supported parameters
 */
export default class ProductsRequest extends AmazonMwsRequest {
  constructor(action, params) {
    const opts = {
      name: 'Products',
      group: 'Products',
      path: '/Products/2011-10-01',
      version: '2011-10-01',
      legacy: false,
      action,
      params,
    };
    super(opts);
  }
}
/**
 * Ojects to represent enum collections used by some request(s)
 * @type {Object}
 */
const enums = exports.enums = {

  ItemConditions() {
    return new EnumType(['New', 'Used', 'Collectible', 'Refurbished', 'Club']);
  },

};

/**
 * Contains brief definitions for unique data type values.
 * Can be used to explain input/output to users via tooltips, for example
 * @type {Object}
 */
const types = exports.types = {

  CompetitivePriceId: {
    1: 'New Buy Box Price',
    2: 'Used Buy Box Price',
  },

  ServiceStatus: {
    GREEN: 'The service is operating normally.',
    GREEN_I: 'The service is operating normally + additional info provided',
    YELLOW: 'The service is experiencing higher than normal error rates or degraded performance.',
    RED: 'The service is unabailable or experiencing extremely high error rates.',
  },

};

/**
 * A collection of currently supported request constructors. Once created and
 * configured, the returned requests can be passed to an mws client `invoke` call
 * @type {Object}
 */
const calls = exports.requests = {

  /**
   * Requests the operational status of the Products API section.
   */
  GetServiceStatus() {
    return new ProductsRequest('GetServiceStatus', {});
  },

  /**
   * Returns a list of products and their attributes, ordered by relevancy,
   * based on a search query that you specify
   */
  ListMatchingProducts() {
    return new ProductsRequest('ListMatchingProducts', {
      MarketplaceId: { name: 'MarketplaceId', required: true },
      Query: { name: 'Query', required: true },
      QueryContextId: { name: 'QueryContextId' },
    });
  },

  /**
   * Returns a list of products and their attributes,
   * based on a list of ASIN values that you specify
   */
  GetMatchingProduct() {
    return new ProductsRequest('GetMatchingProduct', {
      MarketplaceId: { name: 'MarketplaceId', required: true },
      ASINList: { name: 'ASINList.ASIN', list: true, required: true },
    });
  },

  /**
   * Returns a list of products and their attributes,
   * based on a list of specified ID values that you specify
   */
  GetMatchingProductForId() {
    return new ProductsRequest('GetMatchingProductForId', {
      MarketplaceId: { name: 'MarketplaceId', required: true },
      IdType: { name: 'IdType', required: true },
      IdList: { name: 'IdList.Id', list: true, required: true },
    });
  },

  /**
   * Returns the current competitive pricing of a product,
   * based on the SellerSKU and MarketplaceId that you specify
   */
  GetCompetitivePricingForSKU() {
    return new ProductsRequest('GetCompetitivePricingForSKU', {
      MarketplaceId: { name: 'MarketplaceId', required: true },
      SellerSKUList: { name: 'SellerSKUList.SellerSKU', list: true, required: true },
    });
  },

  /**
   * Same as above, except that it uses a MarketplaceId and an ASIN to uniquely
   * identify a product, and it does not return the SKUIdentifier element
   */
  GetCompetitivePricingForASIN() {
    return new ProductsRequest('GetCompetitivePricingForASIN', {
      MarketplaceId: { name: 'MarketplaceId', required: true },
      ASINList: { name: 'ASINList.ASIN', list: true, required: true },
    });
  },

  /**
   * Returns the lowest price offer listings for a specific product by item condition.
   */
  GetLowestOfferListingsForSKU() {
    return new ProductsRequest('GetLowestOfferListingsForSKU', {
      MarketplaceId: { name: 'MarketplaceId', required: true },
      ItemCondition: { name: 'ItemCondition' },
      SellerSKUList: { name: 'SellerSKUList.SellerSKU', list: true, required: true },
    });
  },

  /**
   * Same as above but by a list of ASIN's you provide
   */
  GetLowestOfferListingsForASIN() {
    return new ProductsRequest('GetLowestOfferListingsForASIN', {
      MarketplaceId: { name: 'MarketplaceId', required: true },
      ItemCondition: { name: 'ItemCondition' },
      ASINList: { name: 'ASINList.ASIN', list: true, required: true },
    });
  },

  /**
   * Returns the product categories that a product belongs to,
   * including parent categories back to the root for the marketplace
   */
  GetProductCategoriesForSKU() {
    return new ProductsRequest('GetProductCategoriesForSKU', {
      MarketplaceId: { name: 'MarketplaceId', required: true },
      SellerSKU: { name: 'SellerSKU', required: true },
    });
  },

  /**
   * Same as above, except that it uses a MarketplaceId and an ASIN to
   * uniquely identify a product.
   */
  GetProductCategoriesForASIN() {
    return new ProductsRequest('GetProductCategoriesForASIN', {
      MarketplaceId: { name: 'MarketplaceId', required: true },
      ASIN: { name: 'ASIN', required: true },
    });
  },

  /**
   * Returns pricing information for your own offer listings, based on ASIN.
   *
   */
  GetMyPriceForASIN() {
    return new ProductsRequest('GetMyPriceForASIN', {
      MarketplaceId: { name: 'MarketplaceId', required: true },
      ASINList: { name: 'ASINList.ASIN', list: true, required: true },
    });
  },

  /**
   * Returns pricing information for your own offer listings,
   * based on SellerSKU.
   */
  GetMyPriceForSKU() {
    return new ProductsRequest('GetMyPriceForSKU', {
      MarketplaceId: { name: 'MarketplaceId', required: true },
      SellerSKUList: { name: 'SellerSKUList.SellerSKU', list: true, required: true },
    });
  },
};
