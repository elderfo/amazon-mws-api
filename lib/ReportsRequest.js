/**
 * Reports API requests and definitions for Amazon's MWS web services.
 * For information on using, please see examples folder.
 *
 * @author Robert Saunders
 */
import AmazonMwsRequest from './AmazonMwsRequest';
import EnumType from './EnumType';


/**
 * Construct a Reports API request for mws.Client.invoke()
 *
 * @param {String} action Action parameter of request
 * @param {Object} params Schemas for all supported parameters
 */
export default class ReportsRequest extends AmazonMwsRequest {
  constructor(action, params) {
    const opts = {
      name: 'Reports',
      group: 'Reports & Report Scheduling',
      path: '/',
      version: '2009-01-01',
      legacy: true,
      action,
      params,
    };
    super(opts);
  }
}

/**
 * Ojects to represent enum collections used by some request(s)
 * @type {Object}
 */
export const enums = {

  Schedules() {
    return new EnumType(['_15_MINUTES_', '_30_MINUTES_', '_1_HOUR_', '_2_HOURS_', '_4_HOURS_', '_8_HOURS_', '_12_HOURS_', '_72_HOURS_', '_1_DAY_', '_2_DAYS_', '_7_DAYS_', '_14_DAYS_', '_15_DAYS_', '_30_DAYS_', '_NEVER_']);
  },

  ReportProcessingStatuses() {
    return new EnumType(['_SUBMITTED_', '_IN_PROGRESS_', '_CANCELLED_', '_DONE_', '_DONE_NO_DATA_']);
  },

  ReportOptions() {
    return new EnumType(['ShowSalesChannel=true']);
  },

};

/**
 * A collection of currently supported request constructors. Once created and
 * configured, the returned requests can be passed to an mws client `invoke` call
 * @type {Object}
 */
export const requests = {

  GetReport() {
    return new ReportsRequest('GetReport', {
      ReportId: { name: 'ReportId', required: true },
    });
  },

  GetReportCount() {
    return new ReportsRequest('GetReportCount', {
      ReportTypeList: { name: 'ReportTypeList.Type', list: true },
      Acknowledged: { name: 'Acknowledged', type: 'Boolean' },
      AvailableFromDate: { name: 'AvailableFromDate', type: 'Timestamp' },
      AvailableToDate: { name: 'AvailableToDate', type: 'Timestamp' },
    });
  },

  GetReportList() {
    return new ReportsRequest('GetReportList', {
      MaxCount: { name: 'MaxCount' },
      ReportTypeList: { name: 'ReportTypeList.Type', list: true },
      Acknowledged: { name: 'Acknowledged', type: 'Boolean' },
      AvailableFromDate: { name: 'AvailableFromDate', type: 'Timestamp' },
      AvailableToDate: { name: 'AvailableToDate', type: 'Timestamp' },
      ReportRequestIdList: { name: 'ReportRequestIdList.Id', list: true },
    });
  },

  GetReportListByNextToken() {
    return new ReportsRequest('GetReportListByNextToken', {
      NextToken: { name: 'NextToken', required: true },
    });
  },

  GetReportRequestCount() {
    return new ReportsRequest('GetReportRequestCount', {
      RequestedFromDate: { name: 'RequestedFromDate', type: 'Timestamp' },
      RequestedToDate: { name: 'RequestedToDate', type: 'Timestamp' },
      ReportTypeList: { name: 'ReportTypeList.Type', list: true },
      ReportProcessingStatusList: {
        name: 'ReportProcessingStatusList.Status',
        list: true,
        type: 'reports.ReportProcessingStatuses',
      },
    });
  },

  GetReportRequestList() {
    return new ReportsRequest('GetReportRequestList', {
      MaxCount: { name: 'MaxCount' },
      RequestedFromDate: { name: 'RequestedFromDate', type: 'Timestamp' },
      RequestedToDate: { name: 'RequestedToDate', type: 'Timestamp' },
      ReportRequestIdList: { name: 'ReportRequestIdList.Id', list: true },
      ReportTypeList: { name: 'ReportTypeList.Type', list: true },
      ReportProcessingStatuses: {
        name: 'ReportProcessingStatusList.Status',
        list: true,
        type: 'reports.ReportProcessingStatuses',
      },
    });
  },

  GetReportRequestListByNextToken() {
    return new ReportsRequest('GetReportRequestListByNextToken', {
      NextToken: { name: 'NextToken', required: true },
    });
  },

  CancelReportRequests() {
    return new ReportsRequest('CancelReportRequests', {
      RequestedFromDate: { name: 'RequestedFromDate', type: 'Timestamp' },
      RequestedToDate: { name: 'RequestedToDate', type: 'Timestamp' },
      ReportRequestIdList: { name: 'ReportRequestIdList.Id', list: true },
      ReportTypeList: { name: 'ReportTypeList.Type', list: true },
      ReportProcessingStatusList: {
        name: 'ReportProcessingStatusList.Status',
        list: true,
        type: 'reports.ReportProcessingStatuses',
      },
    });
  },

  RequestReport() {
    return new ReportsRequest('RequestReport', {
      ReportType: { name: 'ReportType', required: true },
      MarketplaceIdList: { name: 'MarketplaceIdList.Id', list: true, required: false },
      StartDate: { name: 'StartDate', type: 'Timestamp' },
      EndDate: { name: 'EndDate', type: 'Timestamp' },
      ReportOptions: { name: 'ReportOptions', type: 'reports.ReportOptions' },
    });
  },

  ManageReportSchedule() {
    return new ReportsRequest('ManageReportSchedule', {
      ReportType: { name: 'ReportType', required: true },
      Schedule: { name: 'Schedule', type: 'reports.Schedules', required: true },
      ScheduleDate: { name: 'ScheduleDate', type: 'Timestamp' },
    });
  },

  GetReportScheduleList() {
    return new ReportsRequest('GetReportScheduleList', {
      ReportTypeList: { name: 'ReportTypeList.Type', list: true },
    });
  },

  GetReportScheduleListByNextToken() {
    return new ReportsRequest('GetReportScheduleListByNextToken', {
      NextToken: { name: 'NextToken', required: true },
    });
  },

  GetReportScheduleCount() {
    return new ReportsRequest('GetReportScheduleCount', {
      ReportTypeList: { name: 'ReportTypeList.Type', list: true },
    });
  },

  UpdateReportAcknowledgements() {
    return new ReportsRequest('UpdateReportAcknowledgements', {
      ReportIdList: { name: 'ReportIdList.Id', list: true, required: true },
      Acknowledged: { name: 'Acknowledged', type: 'Boolean' },
    });
  },

};
