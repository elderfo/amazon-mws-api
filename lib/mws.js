

import AmazonMwsRequest from './AmazonMwsRequest';
import EnumType from './EnumType';
import ComplexListType from './ComplexListType';
import AmazonMwsClient from './AmazonMwsClient';

import * as FulfillmentRequest from './FulfillmentRequest';
import * as OrdersRequest from './OrdersRequest';
import * as SellersRequest from './SellersRequest';
import * as FeedsRequest from './FeedsRequest';
import * as ProductsRequest from './ProductsRequest';
import * as ReportsRequest from './ReportsRequest';

exports.Client = AmazonMwsClient;
exports.Request = AmazonMwsRequest;
exports.Enum = EnumType;
exports.ComplexList = ComplexListType;
exports.Fbs = FulfillmentRequest;
exports.Orders = OrdersRequest;
exports.Sellers = SellersRequest;
exports.Feeds = FeedsRequest;
exports.Products = ProductsRequest;
exports.Reports = ReportsRequest;
