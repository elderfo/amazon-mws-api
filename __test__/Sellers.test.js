import { Sellers } from '../lib/mws';

describe('SellersRequest', () => {
  it('should not be null', () => {
    expect(Sellers).toBeDefined();
  });

  it('should have the expected calls', () => {
    const expectedCalls = ['GetServiceStatus', 'ListMarketplaceParticipations', 'ListMarketplaceParticipationsByNextToken'];
    const actualKeys = Object.keys(Sellers.requests);

    expect(actualKeys).toEqual(expectedCalls);
  });

  test('GetServiceStatus should populate properties', () => {
    const request = Sellers.requests.GetServiceStatus();
    expect(request).toMatchSnapshot();
  });

  test('ListMarketplaceParticipations should populate properties', () => {
    const request = Sellers.requests.ListMarketplaceParticipations();
    expect(request).toMatchSnapshot();
  });

  test('ListMarketplaceParticipationsByNextToken should populate properties', () => {
    const request = Sellers.requests.ListMarketplaceParticipationsByNextToken();
    expect(request).toMatchSnapshot();
  });

  test('types should return expected values', () => {
    const types = Sellers.types;
    expect(types).toMatchSnapshot();
  });
});
