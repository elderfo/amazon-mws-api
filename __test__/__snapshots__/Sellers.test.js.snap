// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SellersRequest GetServiceStatus should populate properties 1`] = `
SellersRequest {
  "action": "GetServiceStatus",
  "api": Object {
    "legacy": false,
    "path": "/Sellers/2011-07-01",
    "upload": undefined,
    "version": "2011-07-01",
  },
  "params": Object {},
  "paramsMap": Object {},
}
`;

exports[`SellersRequest ListMarketplaceParticipations should populate properties 1`] = `
SellersRequest {
  "action": "ListMarketplaceParticipations",
  "api": Object {
    "legacy": false,
    "path": "/Sellers/2011-07-01",
    "upload": undefined,
    "version": "2011-07-01",
  },
  "params": Object {},
  "paramsMap": Object {},
}
`;

exports[`SellersRequest ListMarketplaceParticipationsByNextToken should populate properties 1`] = `
SellersRequest {
  "action": "ListMarketplaceParticipationsByNextToken",
  "api": Object {
    "legacy": false,
    "path": "/Sellers/2011-07-01",
    "upload": undefined,
    "version": "2011-07-01",
  },
  "params": Object {
    "NextToken": Object {
      "name": "NextToken",
      "required": true,
    },
  },
  "paramsMap": Object {},
}
`;

exports[`SellersRequest types should return expected values 1`] = `
Object {
  "ServiceStatus": Object {
    "GREEN": "The service is operating normally.",
    "GREEN_I": "The service is operating normally + additional info provided",
    "RED": "The service is unabailable or experiencing extremely high error rates.",
    "YELLOW": "The service is experiencing higher than normal error rates or degraded performance.",
  },
}
`;
